src/platforms/hosted/windows/ftdi.c

```diff
@@ -0,0 +1,188 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "general.h"
+#include "ftdi.h"
+#include "ftd2xx.h"
+#include "bmp_hosted.h"
+
+FT_HANDLE ftHandle;
+
+/*
+	The following structure is mocked by this module. FTD2XX
+	does not provide a similar structure.
+	
+	Values mocked:
+		ftdi_ctx.type	- The type of FTDI chip in the adapter
+*/
+static ftdi_context_s ftdi_ctx = {0}; // Used to fake libusb context and pass required parameters back to caller
+
+/*
+	The following array is used to map FTD2XX device type identifiers
+	to libftdi identifiers. The array is ordered by the FTD2XX
+	values, with the array entries being the libftdi values
+*/
+
+int ftdi_chip_types[] = {TYPE_AM, TYPE_BM,
+	-1, // FT_DEVICE_100AX not supported
+	-2, // Unknown type
+	TYPE_2232C, TYPE_R, TYPE_2232H, TYPE_4232H, TYPE_232H, TYPE_230X};
+
+#define number_of_ftdi_chip_types (int)(sizeof(ftdi_chip_types) / sizeof(int))
+
+#define READ_TIMEOUT  500 // Expressed in milliseconds
+#define WRITE_TIMEOUT 500
+
+struct ftdi_context *ftdi_new(void)
+{
+	return &ftdi_ctx; // JUst need to fake the structure being created
+}
+
+int ftdi_set_interface(struct ftdi_context *ftdi, enum ftdi_interface the_interface)
+{
+	(void)ftdi;
+	FT_STATUS ftStatus;
+	FT_DEVICE ftDevice;
+	DWORD device_id;
+	char serialNumber[16] = {0};
+	char description[64];
+	//
+	// FTD2XX needs a qualified serial number to open the correct device. Append
+	// an interface letter to the serial number by adding the number to 'A'
+	//
+	strcpy(serialNumber, info.serial);
+	serialNumber[strlen(serialNumber)] = 'A' + (the_interface - 1);
+
+	ftStatus = FT_OpenEx(serialNumber, FT_OPEN_BY_SERIAL_NUMBER, &ftHandle);
+	if (ftStatus != FT_OK)
+		return 0;
+	ftStatus = FT_SetTimeouts(ftHandle, READ_TIMEOUT, WRITE_TIMEOUT);
+	if (ftStatus != FT_OK)
+		return 0;
+	ftStatus = FT_GetDeviceInfo(ftHandle, &ftDevice, &device_id, serialNumber, description, NULL);
+	if (ftStatus == FT_OK) {
+		int device_type_index = ftDevice;
+		if (device_type_index <= number_of_ftdi_chip_types) {
+			ftdi_ctx.type = ftdi_chip_types[device_type_index];
+		}
+	}
+	return 0;
+}
+
+int ftdi_usb_open_desc(struct ftdi_context *ftdi, int vendor, int product, const char *description, const char *serial)
+{
+	(void)ftdi;
+	(void)vendor;
+	(void)product;
+	(void)description;
+	(void)serial;
+
+	return 0;
+}
+
+int ftdi_usb_close(struct ftdi_context *ftdi)
+{
+	FT_STATUS ftStatus;
+	int result;
+	(void)ftdi;
+	ftStatus = FT_Close(ftHandle);
+	if (ftStatus == FT_OK)
+		result = 0;
+	else
+		result = -1;
+	return result;
+}
+
+void ftdi_free(struct ftdi_context *ftdi)
+{
+	(void)ftdi;
+}
+
+int ftdi_set_baudrate(struct ftdi_context *ftdi, int baudrate)
+{
+	(void)ftdi;
+
```

Comment:
Please remove this extra newline (same comment for ftdi_set_bitmode, ftdi_usb_purge_buffers, and the last two functions in the file)